stan/math/prim/functor/multi_expression.hpp

```diff
@@ -0,0 +1,196 @@
+#ifndef STAN_MATH_PRIM_FUNCTOR_MULTI_EXPRESSION_HPP
+#define STAN_MATH_PRIM_FUNCTOR_MULTI_EXPRESSION_HPP
+
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/err.hpp>
+#include <tuple>
+
+namespace stan {
+namespace math {
+
+namespace internal {
+constexpr int static_sum() { return 0; }
+
+template <typename Arg0, typename... Args>
+constexpr int static_sum(Arg0 arg0, Args... args) {
+  return arg0 + static_sum(args...);
+}
+
+}  // namespace internal
+
+/**
+ * Represents multiple exprs that will be calculated in same loop.
+ * @tparam T_expressions types of exprs
+ */
+template <typename... T_expressions>
+class eigen_expressions_ {
+ public:
+  /**
+   * Constructor.
+   * @param exprs expresions that will be calculated in same loop.
+   * @throw invalid_argument expressions have different sizes (for row major
+   * expressions rows and columns are swapped for the check)
+   */
+  explicit eigen_expressions_(T_expressions&&... exprs)
+      : exprs_(std::forward<T_expressions>(exprs)...) {
+    index_apply<sizeof...(T_expressions) - 1>([&](auto... Is) {
+      constexpr auto first_flags = Eigen::internal::traits<std::decay_t<
+          std::tuple_element_t<0, std::tuple<T_expressions...>>>>::Flags;
+      static_cast<void>(std::initializer_list<int>{
+          ((((Eigen::internal::traits<std::decay_t<std::tuple_element_t<
+                  Is + 1, std::tuple<T_expressions...>>>>::Flags
+              ^ first_flags)
+             & Eigen::RowMajorBit)
+                ? check_matching_dims("eigen_expressions_.eigen_expressions_",
+                                      "first expression", std::get<0>(exprs_),
+                                      "transposed expression",
+                                      std::get<Is + 1>(exprs_).transpose())
+                : check_matching_dims("eigen_expressions_.eigen_expressions_",
+                                      "first expression", std::get<0>(exprs_),
+                                      "expression", std::get<Is + 1>(exprs_))),
+           0)...});
+    });
+  }
+
+ private:
+  std::tuple<T_expressions...> exprs_;
+  template <typename... T_results>
+  friend class eigen_results_;
+};
+
+/**
+ * Deduces types for constructing \c expressions_ object.
+ * @tparam T_expressions types of exprs
+ * @param exprs exprs that will be used in same kernel.
+ */
+template <typename... T_expressions,
+          require_all_eigen_t<T_expressions...>* = nullptr>
+eigen_expressions_<T_expressions...> eigen_expressions(
+    T_expressions&&... exprs) {
+  return eigen_expressions_<T_expressions...>(
+      std::forward<T_expressions>(exprs)...);
+}
+
+/**
+ * Represents results that will be calculated in same loop.
+ * @tparam T_results types of results
+ */
+template <typename... T_results>
+class eigen_results_ {
+  std::tuple<T_results...> results_;
```

Comment:
Should this object's tuple hold actual types or just references?

Reply:
It uses perfect forwarding - for rvalues it holds objects and for lvalues it holds references.